Add configuration option `fail_on_warnings`

Dschee opened the pull request:
This is a naive fix for #22. I tested this by installing my fork via Mint and it worked. I'm not sure if this is complete though, as I didn't write any tests or check if this also works as command line argument directly (I just tested via configuration file).

ileitch commented on Sources/PeripheryKit/Configuration.swift:
```diff
@@ -28,6 +28,7 @@ public class Configuration: Singleton {
     var aggressive: Bool = false
     var updateCheck: Bool = true
     var diagnosisConsole: Bool = false
+    var failOnWarnings: Bool = false
```
I think just `strict` would be a better name. The results are only printed as warnings for the Xcode formatter, so it'd be nice for this to be more general. Plus `strict` is quite common in other tools.

Dschee replied:
Renamed to `strict`.

ileitch commented on Sources/PeripheryKit/Frontend/Commands/ScanCommand.swift:
```diff
@@ -221,6 +223,10 @@ public struct ScanOptions: OptionsProtocol {
+
+            <*> mode <| Option(key: "fail-on-warnings",
+                               defaultValue: BoolValue(config.failOnWarnings),
+                               usage: "Make sure command fails if any warnings are encountered")
```
Change this to "Exit with non-zero status if any unused code is found"

Dschee replied:
Done.

ileitch commented on Sources/PeripheryKit/Frontend/Commands/ScanCommand.swift:
```diff
@@ -221,6 +223,10 @@ public struct ScanOptions: OptionsProtocol {
                                defaultValue: BoolValue(!config.updateCheck),
                                usage: "Disable checking for updates")
+
+            <*> mode <| Option(key: "fail-on-warnings",
```
`strict`

Dschee replied:
Fixed.

ileitch commented on Sources/PeripheryKit/Frontend/Commands/ScanBehavior.swift:
```diff
@@ -84,6 +84,10 @@ class ScanBehavior {
+
+            if !filteredDeclarations.isEmpty && configuration.failOnWarnings {
+                throw PeripheryKitError.foundIssues(count: filteredDeclarations.count)
+            }
```
There's a call to `updateChecker.notifyIfAvailable()` further below, let's move it to above this condition so that people still get upgrade notices when using strict mode.

Dschee replied:
Makes sense, fixed.

ileitch commented:
Awesome thanks, could you add a changelog entry too?

Dschee commented:
I'm testing the command line argument and will add a CHANGELOG entry if no issues arise. Stay tuned ...

Dschee commented:
Do the command line arguments matter if there's a configuration file already? Cause it didn't work in my project when running `periphery scan --strict` with this configuration file:

```yaml
project: NewProjectTemplate.xcodeproj
schemes: [App]
targets: [App, Tests, UITests]
report_exclude:
  - App/Generated/SwiftGen/*.swift
  - App/SupportingFiles/*.swift
retain_public: false
quiet: true

```

ileitch commented:
Ah yes, you need to populate the configuration for command-line options. You'll need to do this in both `ScanCommand` and `ScanSyntaxCommand`, e.g:

https://github.com/peripheryapp/periphery/blob/master/Sources/PeripheryKit/Frontend/Commands/ScanCommand.swift#L48

Dschee commented:
Makes sense, I added it as suggested to both. Testing now ...

Please note that I already added a CHANGELOG entry and used the conventions from the SwiftGen project (see their [CHANGELOG](https://github.com/SwiftGen/SwiftGen/blob/master/CHANGELOG.md)) and the SwiftLint project ([theirs](https://github.com/realm/SwiftLint/blob/master/CHANGELOG.md)) here, too since there's none yet. Feel free to change the structure if you want to use a different style, I just thought since this open sourced very recently you might not haved thought about one yet.

Dschee commented:
Yes, it works both with `periphery scan --strict` and `periphery scan-syntax --strict .`. Should be ready to merge! 🎉

ileitch commented:
CHANGELOG format looks good 👍

Dschee commented:
@ileitch Thanks for reviewing and merging so quickly. 👍 

Unfortunately the CHANGELOG doesn't seem to render like in the linked documents, where everything is on one line, but here it usses mutiple lines ... dunno why, but can be fixed over time, shouldn't be a huge problem. 😅

ileitch commented:
Yep no worries, thanks for the submission!
